docs/rfcs/rfc-034-auth.md

```diff
@@ -0,0 +1,859 @@
+```
+
+The arguments have the following utility:
+
+* `enabled` is `true` by default, but can be set to `false` if for instance, you want to disable global authentication for a particular type
+* `operations` is used to specify which operations require authentication, and it specifies all by default
+
+The operations for authentication are not as fine-grained as for authorization, and they can be seen as linking directly to root-level operations, rather than nested operations such as `connect` and `disconnect`.
+
+The directive can be applied to individual objects, interfaces and fields, but can also be applied as a schema extension to enable global authentication.
+
+#### PROPOSAL: JWT payload specification
+
+Allow users to provide a schema describing what their JWT payload looks like. For example, this could be a JSON schema passed into the auth configuration:
+
+```jsonschema
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "type": "object",
+  "properties": {
+    "sub": {
+      "type": "string"
+    },
+    "roles": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+    }
+  },
+  "required": [ "sub" ]
+}
+```
+
+#### PROPOSAL: Access control
+
+Sitting between authentication and authorization evaluation, access control rules will control who can access which top-level operations, as assessed at the beginning of resolver execution.
+
+```gql
+input StringWhere {
+  OR: [StringWhere!]
+  AND: [StringWhere!]
+  NOT: StringWhere
+  equals: String
+  in: [String!]
+  matches: String
+  contains: String
+  startsWith: String
+  endsWith: String
+}
+
+input StringListWhere {
+  OR: [StringListWhere!]
+  AND: [StringListWhere!]
+  NOT: StringListWhere
+  all: [String!]
+  some: [String!]
+  single: String
+}
+
+input JWTPayloadWhere {
+  OR: [JWTPayloadWhere!]
+  AND: [JWTPayloadWhere!]
+  NOT: JWTPayloadWhere
+  sub: StringWhere
+  roles: StringListWhere
+}
+
+input AccessControlWhere {
+  OR: [AccessControlWhere!]
+  AND: [AccessControlWhere!]
+  NOT: AccessControlWhere
+  jwtPayload: JWTPayloadWhere
+}
+
+enum AccessControlOperation {
+  READ
+  CREATE
+  UPDATE
+  DELETE
+  SUBSCRIBE
+}
+
+input AccessControlAllowRule {
+  operations: [AccessControlOperation!]! = [READ, CREATE, UPDATE, DELETE, SUBSCRIBE]
+  requireAuthentication: Boolean! = true
+  where: AccessControlWhere!
+}
+
+directive @accessControl(
+  allow: [AccessControlAllowRule!]
+) on OBJECT | FIELD_DEFINITION | INTERFACE | SCHEMA
+```
+
+The proposal is that this will _only_ be in relation to top-level operations, and not when performing a nested operation on a type. Authorization rules as described below should be used for this use case.
+
+Giving an example of using the above, to prevent a `User` type having its delete Mutation called by anyone but an administrator, one could do:
+
+```gql
+type User @accessControl(allow: [{ operations: [DELETE], where: { jwtPayload: { roles: { single: "ADMIN" } } } }]) {
```

Comment:
What are the limitations that block us to have it as part of the `@authorization` directive?

Reply:
No limitations, apart from I think it will be confusing because this is assessed only once, on entry to the relevant resolver. Authorization is assessed in a nested manner, whenever you hit a specified. I have actually added an alternative suggestion for how we might do this with limited input ourselves, based on what you said earlier!